lib/py/src/transport/TSocket.py

```diff
@@ -74,7 +74,30 @@ def setHandle(self, h):
         self.handle = h
 
     def isOpen(self):
-        return self.handle is not None
+        if self.handle is None:
+            return False
+
+        # this lets us cheaply see if the other end of the socket is still
+        # connected. if disconnected, we'll get EOF back (expressed as zero
+        # bytes of data) otherwise we'll get one byte or an error indicating
+        # we'd have to block for data.
+        #
+        # note that we're not doing this with socket.MSG_DONTWAIT because 1)
+        # it's linux-specific and 2) gevent-patched sockets hide EAGAIN from us
+        # when timeout is non-zero.
+        original_timeout = self.handle.gettimeout()
+        try:
+            self.handle.settimeout(0)
+            try:
+                peeked_bytes = self.handle.recv(1, socket.MSG_PEEK)
+            except (socket.error, OSError) as exc:  # on modern python this is just BlockingIOError
+                if exc.errno == errno.EWOULDBLOCK:
```

Comment:
I think we actually need to check both EWOULDBLOCK and EAGAIN here? From the man page of `recv`:

```
       EAGAIN or EWOULDBLOCK                                                                                                                                                                                                                                                                                                                                                                                                                                          
              The socket is marked nonblocking and the receive operation would block, or a receive timeout had been set and the timeout expired before data was received.  POSIX.1 allows either error to be returned for this case, and does not require these constants to have the same value, so a portable application should check for both possibilities. 
```

Also should EOF be handled explicitly (return False) instead of just raise again?

Reply:
Yeah, seems like it:

```
If no messages are available at the socket [...] the value -1 is
returned and the external variable errno is set to EAGAIN or EWOULDBLOCK
```

Though I don't know what the behaviour is on Windows.

For EOF, returning is better since `isOpen` didn't throw before and it would be a breaking change.

Reply:
Ah right, I forgot the old implementation of isOpen is just a simple check that would never raise.

In that case, I think we need to `return False` explicitly on all other errors, not only EOF.

Reply:
Sounds good!

Reply:
I'll add both and make it swallow the unexpected exception.

FWIW, looks like the most recent OS to have EAGAIN != EWOULDBLOCK  is quite old: https://stackoverflow.com/questions/7003234/which-systems-define-eagain-and-ewouldblock-as-different-values/7003379#7003379 but it makes sense to be most portable here.